### DISCUSSION TITLE
Support of whisper params

### DISCUSSION THREAD

**Discussion by pminev** [alpaca-core/alpaca-core-sdk#58] [Dev]
All params for the inference can be checked [here](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L461): 

## Whisper params needed for the MVP:

- translate -> translate the audio - seems like a reasonable feature
- no_context - do not use past transcription as initial prompt for the decoder
- no_timestamps - no timestamps generation
- single_segment - force single segment
- [initial prompt](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L497-L503)
- [language auto-detection](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L505)
- new_segment_callback -> called for every newly generated text segment. We'll need this if we want to stream the result and not to wait until it's completed.
 - progress_callback -> called on each progress update

## Whisper params not needed for the MVP:

> **_NOTE:_**  All of the following will have valid default values, but will be not exposed to the clients

- params to enable/disable prints on stdout of the internal library during decoding. I don't think they will be useful by any means for the clients and we should hide them
  - print_special;     // print special tokens (e.g. <SOT>, <EOT>, <BEG>, etc.)
  - print_progress;    // print progress information
  - print_realtime;    // print results from within whisper.cpp (avoid it, use callback instead)
  - print_timestamps;  // print timestamps for each text segment when printing realtime

- suppress_regex - [desc](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L494)
- common decoding params:
  - suppress_blank - suppresses blank outputs
  - suppress_non_speech_tokens
  - temperature - initial decoding temp
  - max_initial_ts - [desc](https://github.com/openai/whisper/blob/f82bc59f5ea234d4b97fb2860842ed38519f7e65/whisper/decoding.py#L97)
  - length_penalty - [desc](https://github.com/openai/whisper/blob/f82bc59f5ea234d4b97fb2860842ed38519f7e65/whisper/transcribe.py#L267)

- fallback parameters - seems like they are for fine tuning
  - temperature_inc
  - entropy_thold
  - logprob_thold
  - no_speech_thold - not implemented In whisper either

- grammar -> GBNF grammar to guide decoding. At some point we might need it.
  - grammar_rules
  - n_grammar_rules
  - i_start_rule
  - grammar_penalty

- encoder_begin_callback -> called each time before the encoder starts

### Experimental in whisper

- token level timestamps - [check all of them here](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L478)
  - token_timestamps
  - thold_pt
  - thold_ptsum
  - max_len
  - split_on_word
  - max_tokens
- debug_mode - extra info
- [audio_ctx](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L489) 
- [tdrz_enable](https://github.com/alpaca-core/whisper.cpp/blob/6739eb83c3ca5cf40d24c6fe8442a761a1eb6248/include/whisper.h#L491C34-L491C45)

**Comment by pminev** [comment]
IMO experimental parameters are totally out of context for now.

**Comment by iboB** [comment]
Since the code for tokenization used for initial prompts is largely a copy-paste of llama.cpp, I'm wondering whether we can reuse it. I think that introducing a dependency to whisper.cpp on llama.cpp is not a terrible pain.

**Reply by pminev** [reply]
I need to investigate it a little bit further but isn't that too small chunk of code in order to introduce a dependency between the two?

**Comment by iboB** [comment]
> new_segment_callback 
> progress_callback

Callbacks can be useful for the local implementation but we have no way of transferring them through the API as params. The functionality that callbacks cover will have to pass through the progress callback of sessions.
